File: kirun-js/src/engine/function/system/math/RandomInt.ts
import { Schema } from '../../../json/schema/Schema';
import { Event } from '../../../model/Event';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionSignature } from '../../../model/FunctionSignature';
import { Parameter } from '../../../model/Parameter';
import { Namespaces } from '../../../namespaces/Namespaces';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { MapUtil } from '../../../util/MapUtil';
import { AbstractFunction } from '../../AbstractFunction';

export class RandomInt extends AbstractFunction {
    public static readonly MIN_VALUE = 'minValue';

    public static readonly MAX_VALUE = 'maxValue';

    public static readonly VALUE = 'value';

    private static readonly SIGNATURE: FunctionSignature = new FunctionSignature('Random')
        .setParameters(
            MapUtil.of(
                RandomInt.MIN_VALUE,
                Parameter.of(
                    RandomInt.MIN_VALUE,
                    Schema.ofInteger(RandomInt.MIN_VALUE).setDefaultValue(0),
                ),
                RandomInt.MAX_VALUE,
                Parameter.of(
                    RandomInt.MAX_VALUE,
                    Schema.ofInteger(RandomInt.MAX_VALUE).setDefaultValue(2147483647),
                ),
            ),
        )
        .setNamespace(Namespaces.MATH)
        .setEvents(
            new Map<string, Event>([
                Event.outputEventMapEntry(
                    MapUtil.of(RandomInt.VALUE, Schema.ofInteger(RandomInt.VALUE)),
                ),
            ]),
        );

    public getSignature(): FunctionSignature {
        return RandomInt.SIGNATURE;
    }

    protected internalExecute(context: FunctionExecutionParameters): FunctionOutput {
        let min: number = context.getArguments()?.get(RandomInt.MIN_VALUE);

        let max: number = context.getArguments()?.get(RandomInt.MAX_VALUE);

        let num: number = Math.floor(Math.random() * (max - min) + min);

        return new FunctionOutput([EventResult.outputOf(new Map([[RandomInt.VALUE, num]]))]);
    }
}

File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/function/system/math/RandomInt.java
package com.fincity.nocode.kirun.engine.function.system.math;

import static com.fincity.nocode.kirun.engine.namespaces.Namespaces.MATH;

import java.util.List;
import java.util.Map;
import java.util.Random;

import com.fincity.nocode.kirun.engine.function.reactive.AbstractReactiveFunction;
import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.model.Event;
import com.fincity.nocode.kirun.engine.model.EventResult;
import com.fincity.nocode.kirun.engine.model.FunctionOutput;
import com.fincity.nocode.kirun.engine.model.FunctionSignature;
import com.fincity.nocode.kirun.engine.model.Parameter;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonPrimitive;

import reactor.core.publisher.Mono;

public class RandomInt extends AbstractReactiveFunction {

	private static final String MIN_VALUE = "minValue";

	private static final String MAX_VALUE = "maxValue";

	private static final String VALUE = "value";

	private Random rand = new Random();

	@Override
	public FunctionSignature getSignature() {
		return new FunctionSignature().setName("RandomInt")
		        .setNamespace(MATH)
		        .setParameters(Map.of(MIN_VALUE, new Parameter().setParameterName(MIN_VALUE)
		                .setSchema(Schema.ofInteger(MIN_VALUE)
		                        .setDefaultValue(new JsonPrimitive(0))),
		                MAX_VALUE, new Parameter().setParameterName(MAX_VALUE)
		                        .setSchema(Schema.ofInteger(MAX_VALUE)
		                                .setDefaultValue(new JsonPrimitive(Integer.MAX_VALUE)))))
		        .setEvents(Map.ofEntries(Event.outputEventMapEntry(Map.of(VALUE, Schema.ofInteger(VALUE)))));
	}

	@Override
	protected Mono<FunctionOutput> internalExecute(ReactiveFunctionExecutionParameters context) {

		int minValue = context.getArguments()
		        .get(MIN_VALUE)
		        .getAsInt();

		int maxValue = context.getArguments()
		        .get(MAX_VALUE)
		        .getAsInt();

		int randValue = rand.nextInt(minValue, maxValue == Integer.MAX_VALUE ? maxValue : maxValue + 1);

		return Mono
		        .just(new FunctionOutput(List.of(EventResult.outputOf(Map.of(VALUE, new JsonPrimitive(randValue))))));
	}

}
